File: apollo-api/src/main/java/com/apollographql/apollo/api/ScalarTypeAdapters.java
package com.apollographql.apollo.api;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.jetbrains.annotations.NotNull;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

public final class ScalarTypeAdapters {
  private static final Map<Class, CustomTypeAdapter> DEFAULT_ADAPTERS = defaultAdapters();
  private final Map<String, CustomTypeAdapter> customAdapters;

  public ScalarTypeAdapters(@NotNull Map<ScalarType, CustomTypeAdapter> customAdapters) {
    Map<ScalarType, CustomTypeAdapter> nonNullcustomAdapters = checkNotNull(customAdapters, "customAdapters == null");
    this.customAdapters = new HashMap<>();
    for (Map.Entry<ScalarType, CustomTypeAdapter> entry : nonNullcustomAdapters.entrySet()) {
      this.customAdapters.put(entry.getKey().typeName(), entry.getValue());
    }
  }

  @SuppressWarnings("unchecked") @NotNull public <T> CustomTypeAdapter<T> adapterFor(@NotNull ScalarType scalarType) {
    checkNotNull(scalarType, "scalarType == null");

    CustomTypeAdapter<T> customTypeAdapter = customAdapters.get(scalarType.typeName());
    if (customTypeAdapter == null) {
      customTypeAdapter = DEFAULT_ADAPTERS.get(scalarType.javaType());
    }

    if (customTypeAdapter == null) {
      throw new IllegalArgumentException(String.format("Can't map GraphQL type: %s to: %s. Did you forget to add "
          + "a custom type adapter?", scalarType.typeName(), scalarType.javaType()));
    }

    return customTypeAdapter;
  }

  private static Map<Class, CustomTypeAdapter> defaultAdapters() {
    Map<Class, CustomTypeAdapter> adapters = new LinkedHashMap<>();
    adapters.put(String.class, new DefaultCustomTypeAdapter<String>() {
      @NotNull @Override public String decode(@NotNull CustomTypeValue value) {
        return value.value.toString();
      }
    });
    adapters.put(Boolean.class, new DefaultCustomTypeAdapter<Boolean>() {
      @NotNull @Override public Boolean decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLBoolean) {
          return (Boolean) value.value;
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Boolean.parseBoolean(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't decode: " + value + " into Boolean");
        }
      }
    });
    adapters.put(Integer.class, new DefaultCustomTypeAdapter<Integer>() {
      @NotNull @Override public Integer decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).intValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Integer.parseInt(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't decode: " + value + " into Integer");
        }
      }
    });
    adapters.put(Long.class, new DefaultCustomTypeAdapter<Long>() {
      @NotNull @Override public Long decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).longValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Long.parseLong(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't decode: " + value + " into Long");
        }
      }
    });
    adapters.put(Float.class, new DefaultCustomTypeAdapter<Float>() {
      @NotNull @Override public Float decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).floatValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Float.parseFloat(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't decode: " + value + " into Float");
        }
      }
    });
    adapters.put(Double.class, new DefaultCustomTypeAdapter<Double>() {
      @NotNull @Override public Double decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLNumber) {
          return ((Number) value.value).doubleValue();
        } else if (value instanceof CustomTypeValue.GraphQLString) {
          return Double.parseDouble(((CustomTypeValue.GraphQLString) value).value);
        } else {
          throw new IllegalArgumentException("Can't decode: " + value + " into Double");
        }
      }
    });
    adapters.put(FileUpload.class, new CustomTypeAdapter<FileUpload>() {
      @Override public FileUpload decode(@NotNull CustomTypeValue value) {
        return null;
      }

      @NotNull
      @Override
      public CustomTypeValue encode(@NotNull FileUpload value) {
        return new CustomTypeValue.GraphQLString(null);
      }
    });
    adapters.put(Object.class, new DefaultCustomTypeAdapter<Object>() {
      @NotNull @Override public Object decode(@NotNull CustomTypeValue value) {
        return value.value;
      }
    });
    adapters.put(Map.class, new DefaultCustomTypeAdapter<Map>() {
      @NotNull @Override public Map decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLJsonObject) {
          return ((CustomTypeValue.GraphQLJsonObject) value).value;
        } else {
          throw new IllegalArgumentException("Can't decode: " + value + " into Map");
        }
      }
    });
    adapters.put(List.class, new DefaultCustomTypeAdapter<List>() {
      @NotNull @Override public List decode(@NotNull CustomTypeValue value) {
        if (value instanceof CustomTypeValue.GraphQLJsonList) {
          return ((CustomTypeValue.GraphQLJsonList) value).value;
        } else {
          throw new IllegalArgumentException("Can't decode: " + value + " into List");
        }
      }
    });
    return adapters;
  }

  private abstract static class DefaultCustomTypeAdapter<T> implements CustomTypeAdapter<T> {
    @NotNull @Override public CustomTypeValue encode(@NotNull T value) {
      return CustomTypeValue.fromRawValue(value);
    }
  }
}

File: apollo-api/src/main/java/com/apollographql/apollo/api/ScalarTypeAdapters.kt
package com.apollographql.apollo.api

import com.apollographql.apollo.api.CustomTypeValue.*
import com.apollographql.apollo.api.CustomTypeValue.Companion.fromRawValue
import java.io.File

class ScalarTypeAdapters(customAdapters: Map<ScalarType, CustomTypeAdapter<*>>) {
  private val customAdapters = customAdapters.mapKeys { it.key.typeName() }

  @Suppress("UNCHECKED_CAST")
  fun <T : Any> adapterFor(scalarType: ScalarType): CustomTypeAdapter<T> {
    var customTypeAdapter: CustomTypeAdapter<*>? = customAdapters[scalarType.typeName()]
    if (customTypeAdapter == null) {
      customTypeAdapter = DEFAULT_ADAPTERS[scalarType.javaType()]
    }
    return requireNotNull(customTypeAdapter) {
      "Can't map GraphQL type: `${scalarType.typeName()}` to: `${scalarType.javaType()}`. Did you forget to add a custom type adapter?"
    } as CustomTypeAdapter<T>
  }

  private abstract class DefaultCustomTypeAdapter<T : Any> : CustomTypeAdapter<T> {
    override fun encode(value: T): CustomTypeValue<*> {
      return fromRawValue(value)
    }
  }

  companion object {
    @JvmStatic
    val DEFAULT = ScalarTypeAdapters(emptyMap())

    private val DEFAULT_ADAPTERS = mapOf(
        String::class.java to object : DefaultCustomTypeAdapter<String>() {
          override fun decode(value: CustomTypeValue<*>): String {
            return value.value.toString()
          }
        },
        java.lang.Boolean::class.java to object : DefaultCustomTypeAdapter<Boolean>() {
          override fun decode(value: CustomTypeValue<*>): Boolean {
            return when (value) {
              is GraphQLBoolean -> value.value
              is GraphQLString -> java.lang.Boolean.parseBoolean(value.value)
              else -> throw IllegalArgumentException("Can't decode: $value into Boolean")
            }
          }
        },
        java.lang.Integer::class.java to object : DefaultCustomTypeAdapter<Int>() {
          override fun decode(value: CustomTypeValue<*>): Int {
            return when (value) {
              is GraphQLNumber -> value.value.toInt()
              is GraphQLString -> value.value.toInt()
              else -> throw IllegalArgumentException("Can't decode: $value into Integer")
            }
          }
        },
        java.lang.Long::class.java to object : DefaultCustomTypeAdapter<Long>() {
          override fun decode(value: CustomTypeValue<*>): Long {
            return when (value) {
              is GraphQLNumber -> value.value.toLong()
              is GraphQLString -> value.value.toLong()
              else -> throw IllegalArgumentException("Can't decode: $value into Long")
            }
          }
        },
        java.lang.Float::class.java to object : DefaultCustomTypeAdapter<Float>() {
          override fun decode(value: CustomTypeValue<*>): Float {
            return when (value) {
              is GraphQLNumber -> value.value.toFloat()
              is GraphQLString -> value.value.toFloat()
              else -> throw IllegalArgumentException("Can't decode: $value into Float")
            }
          }
        },
        java.lang.Double::class.java to object : DefaultCustomTypeAdapter<Double>() {
          override fun decode(value: CustomTypeValue<*>): Double {
            return when (value) {
              is GraphQLNumber -> value.value.toDouble()
              is GraphQLString -> value.value.toDouble()
              else -> throw IllegalArgumentException("Can't decode: $value into Double")
            }
          }
        },
        Boolean::class.java to object : DefaultCustomTypeAdapter<Boolean>() {
          override fun decode(value: CustomTypeValue<*>): Boolean {
            return when (value) {
              is GraphQLBoolean -> value.value
              is GraphQLString -> java.lang.Boolean.parseBoolean(value.value)
              else -> throw IllegalArgumentException("Can't decode: $value into Boolean")
            }
          }
        },
        Int::class.java to object : DefaultCustomTypeAdapter<Int>() {
          override fun decode(value: CustomTypeValue<*>): Int {
            return when (value) {
              is GraphQLNumber -> value.value.toInt()
              is GraphQLString -> value.value.toInt()
              else -> throw IllegalArgumentException("Can't decode: $value into Integer")
            }
          }
        },
        Long::class.java to object : DefaultCustomTypeAdapter<Long>() {
          override fun decode(value: CustomTypeValue<*>): Long {
            return when (value) {
              is GraphQLNumber -> value.value.toLong()
              is GraphQLString -> value.value.toLong()
              else -> throw IllegalArgumentException("Can't decode: $value into Long")
            }
          }
        },
        Float::class.java to object : DefaultCustomTypeAdapter<Float>() {
          override fun decode(value: CustomTypeValue<*>): Float {
            return when (value) {
              is GraphQLNumber -> value.value.toFloat()
              is GraphQLString -> value.value.toFloat()
              else -> throw IllegalArgumentException("Can't decode: $value into Float")
            }
          }
        },
        Double::class.java to object : DefaultCustomTypeAdapter<Double>() {
          override fun decode(value: CustomTypeValue<*>): Double {
            return when (value) {
              is GraphQLNumber -> value.value.toDouble()
              is GraphQLString -> value.value.toDouble()
              else -> throw IllegalArgumentException("Can't decode: $value into Double")
            }
          }
        },
        FileUpload::class.java to object : CustomTypeAdapter<FileUpload> {
          override fun decode(value: CustomTypeValue<*>): FileUpload {
            return FileUpload("", File(""));
          }

          override fun encode(value: FileUpload): CustomTypeValue<*> {
            return GraphQLString(value.mimetype)
          }
        },
        Map::class.java to object : DefaultCustomTypeAdapter<Map<*, *>>() {
          override fun decode(value: CustomTypeValue<*>): Map<*, *> {
            return if (value is GraphQLJsonObject) {
              value.value
            } else {
              throw IllegalArgumentException("Can't decode: $value into Map")
            }
          }
        },
        List::class.java to object : DefaultCustomTypeAdapter<List<*>>() {
          override fun decode(value: CustomTypeValue<*>): List<*> {
            return if (value is GraphQLJsonList) {
              value.value
            } else {
              throw IllegalArgumentException("Can't decode: $value into List")
            }
          }
        },
        Any::class.java to object : DefaultCustomTypeAdapter<Any>() {
          override fun decode(value: CustomTypeValue<*>): Any {
            return value.value!!
          }
        }
    )
  }
}
